cpp_common/common/UDPConnection.cpp

```diff
@@ -0,0 +1,227 @@
+#include "UDPConnection.h"
+#include <exception>
+#include <unistd.h>
+#include <string.h>
+#include <stdio.h>
+#include <time.h>
+#include <errno.h>
+#include <vector>
+#include <chrono>
+
+using namespace std;
+using namespace std::literals::chrono_literals; // BLACK MAGIC!
+
+// Extra magic for Windows.
+static void initSocketLibrary(void) {
+    static bool initialised = false;
+    if (initialised) {
+	return;
+    }
+
+#ifdef WIN32
+    WSADATA wsaData; // if this doesn't work
+    //WSAData wsaData; // then try this instead
+#include <UDPConnection.h>
+    if (WSAStartup(MAKEWORD(1, 1), &wsaData) != 0) {
+        fprintf(stderr, "WSAStartup failed.\n");
+        exit(1);
+    }
+#endif // WIN32
+    initialised = true;
+}
+
+template<class SockType, class TimeRep, class TimePeriod>
+static void setSocketTimeout(
+	SockType sock, chrono::duration<TimeRep, TimePeriod> d)
+{
+#ifdef WIN32
+    auto ms = chrono::duration_cast<chrono::milliseconds>(d);
+    DWORD timeout = (DWORD) ms.count();
+#else
+    auto sec = chrono::duration_cast<chrono::seconds>(d);
+    struct timeval timeout;
+    timeout.tv_sec = sec.count();
+    timeout.tv_usec = chrono::duration_cast<chrono::microseconds>(
+	    d - sec).count();
+#endif
+    int err = setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO,
+	    (const char*) &timeout, sizeof(timeout));
+    if (err < 0) {
+	throw "Socket timeout could not be set";
+    }
+}
+
+UDPConnection::UDPConnection(
+        int local_port,
+        const char *local_host,
+        int remote_port,
+        const char *remote_host)
+    : local_ip_address(htonl(INADDR_ANY))
+{
+    initSocketLibrary();
+    sock = socket(AF_INET, SOCK_DGRAM, 0);
+    if (sock == 0) {
+        throw "Socket could not be created";
+    }
+
+    sockaddr local_address;
+    if (local_host != NULL && local_host[0] != '\0') {
+	local_address = get_address(local_host, local_port);
+    } else {
+	sockaddr_in *local = (sockaddr_in *) &local_address;
+	local->sin_family = AF_INET;
+	local->sin_addr.s_addr = htonl(INADDR_ANY);
+	local->sin_port = htons(local_port);
+    }
+
+    // TODO: What is the correct timeout for a receiver socket?
+    setSocketTimeout(sock, 500ms);
```

Comment:
So… what _is_ the correct timeout for a socket here? (Timeouts just cause the reader loop to go round again, which allows us to detect termination cleanly and costs next to nothing.)